### PR TITLE
Add response statistics

### PR DESCRIPTION
**Pull request by kthui**
Related PRs:
- https://github.com/triton-inference-server/core/pull/325
- https://github.com/triton-inference-server/common/pull/112
- https://github.com/triton-inference-server/square_backend/pull/16

Add tests for response statistics.

### CODE REVIEW

File: qa/L0_response_statistics/response_statistics_test.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        self._min_output_delay_ns = 200000000
+        self._number_of_fail_responses = 2
+        self._number_of_empty_responses = 1
+        self._statistics_counts = []
+        self._grpc_client = grpcclient.InferenceServerClient(
+            "localhost:8001", verbose=True
+        )
+        self._http_client = httpclient.InferenceServerClient("localhost:8000")
+
+    def _generate_streaming_callback_and_response_pair(self):
```
Comment (oandreeva-nv) [inline qa/L0_response_statistics/response_statistics_test.py:50]: A small nit: It may be helpful to have a small description for helper functions (this and later), which briefly describes what they do and what inputs mean

File: qa/L0_response_statistics/response_statistics_test.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        response = []  # [{"result": result, "error": error}, ...]
```
Comment (oandreeva-nv) [inline qa/L0_response_statistics/response_statistics_test.py:51]: nit: put the comment up 1 line

File: qa/L0_response_statistics/test.sh [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+export CUDA_VISIBLE_DEVICES=0
+
+SERVER=/opt/tritonserver/bin/tritonserver
+source ../common/util.sh
+
+RET=0
+
+rm -rf models && mkdir models
+mkdir -p models/square_int32/1 && (cd models/square_int32 && \
+    echo 'name: "square_int32"' >> config.pbtxt && \
```
Comment (oandreeva-nv) [inline qa/L0_response_statistics/test.sh:50]: nit: I think it should work without a name

File: qa/L0_response_statistics/response_statistics_test.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            self._check_statistics_count_and_duration(
+                response_stats, i, "compute_output"
+            )
+            self._check_statistics_count_and_duration(response_stats, i, "success")
+            self._check_statistics_count_and_duration(response_stats, i, "fail")
+            self._check_statistics_count_and_duration(
+                response_stats, i, "empty_response"
+            )
+
+    def test_response_statistics(self):
```
[the remark this answers is not in the document]
Comment (kthui) [reply qa/L0_response_statistics/response_statistics_test.py:170]: Not specific, but we want at least two of them (i.e. two inferences) to make sure the statistics are aggregated.

File: docs/protocol/extension_statistics.md [hunk 5]
```diff
@@ -114,6 +115,12 @@ $model_stat =
   model/version. So, for example, "inference_stats":"success"
   indicates the number of successful inference requests for the model.
 
+- "response_stats" : The aggregate decoupled response statistics for the
+  model/version. For example, { "key" : { "response_stats" : "success" } }
+  indicates the aggregate statistics of successful decoupled response at "key"
+  for the model, where "key" identifies between different decoupled responses
+  generated by the model.
```
Comment (oandreeva-nv) [inline docs/protocol/extension_statistics.md:122]: > The aggregate decoupled response statistics for the model/version.

I think we need to clarify `model/version`. Is it the specific version of a specific model?

>  indicates the aggregate statistics of successful decoupled response at "key" for the model

I am a little bit confused what `at "key" for the model` should mean.

>   where "key" identifies between different decoupled responses generated by the model.

` identifies between different decoupled responses` <- not sure what you mean by this.

File: qa/L0_response_statistics/response_statistics_test.py (hunk 1 above)
Comment (kthui) [reply qa/L0_response_statistics/response_statistics_test.py:51]: Sure. Added.

File: qa/L0_response_statistics/response_statistics_test.py (hunk 2 above)
Comment (kthui) [reply qa/L0_response_statistics/response_statistics_test.py]: Updated.

File: qa/L0_response_statistics/test.sh (hunk 3 above)
Comment (kthui) [reply qa/L0_response_statistics/test.sh]: It should. Name is removed.

File: docs/protocol/extension_statistics.md (hunk 5 above)
Comment (kthui) [reply docs/protocol/extension_statistics.md:122]: > I think we need to clarify model/version. Is it the specific version of a specific model?

It is a specific version of a specific model, because the backend views different versions of a model different models. Let's use "model" for it, because except one or two other places, the rest of the docs refers to "model".

> I am a little bit confused what at "key" for the model should mean.

It identifies between different decoupled responses generated by the model. For example, a model generates 3 responses, the key can be "0", "1" and "2". 

> identifies between different decoupled responses <- not sure what you mean by this.

Basically a response, even an empty one, should be mapped to a uniquely identifiable value -> "key". It is ultimately the backend responsibility for supplying valid keys for grouping response statistics. i.e. send the response statistics at the right order to ensure it is mapped to the correct index ("0", "1", "2", ...).

I added a comment below to ask the user to check with the backend on the set of keys and what each means.

File: qa/L0_response_statistics/response_statistics_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (oandreeva-nv) [reply qa/L0_response_statistics/response_statistics_test.py:184]: We can do it in a separate PR, but a small description for this can help with maintenance

File: docs/protocol/extension_statistics.md [hunk 6]
```diff
@@ -111,8 +112,15 @@ $model_stat =
   DOES NOT include cache hits.
 
 - "inference_stats" : The aggregate statistics for the
-  model/version. So, for example, "inference_stats":"success"
-  indicates the number of successful inference requests for the model.
+  model. So, for example, "inference_stats":"success" indicates the number of
+  successful inference requests for the model.
+
+- "response_stats" : The aggregate decoupled response statistics for the
+  model. For example, { "key" : { "response_stats" : "success" } } indicates the
```
Comment (oandreeva-nv) [inline docs/protocol/extension_statistics.md:119]: Can we do a real example here?  `key` is still confusing. Something like:
Assume there were `N` number of requests sent to server. An example of a possible response statistics is ...,
where <actual value> is a `key`, which helps to differentiate between decoupled responses generated by the model. It is advised ...

**Review by oandreeva-nv — COMMENTED** [review]
Besides readme, this pr lgtm. In case of urgency ping me and I'll approve, we can follow up with clarifications to readme later.

File: qa/L0_response_statistics/response_statistics_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kthui) [reply qa/L0_response_statistics/response_statistics_test.py:184]: Sure, added more description to the test.

File: docs/protocol/extension_statistics.md (hunk 6 above)
Comment (kthui) [reply docs/protocol/extension_statistics.md:119]: Sure, added a brief example for key.

**Review by oandreeva-nv — DISMISSED** [review]
tm,lg

File: docs/protocol/extension_statistics.md [hunk 7]
```diff
@@ -111,8 +112,16 @@ $model_stat =
   DOES NOT include cache hits.
 
 - "inference_stats" : The aggregate statistics for the
-  model/version. So, for example, "inference_stats":"success"
-  indicates the number of successful inference requests for the model.
+  model. So, for example, "inference_stats":"success" indicates the number of
+  successful inference requests for the model.
+
+- "response_stats" : The aggregate decoupled response statistics for the
```
Comment (rmccorm4) [inline docs/protocol/extension_statistics.md:118]:
```suggestion
- "response_stats" : The aggregate response statistics for the
```

Comment (rmccorm4) [reply docs/protocol/extension_statistics.md:118]: would replace mentions of "decoupled response" to just "response"

File: docs/protocol/extension_statistics.md [hunk 8]
```diff
@@ -111,8 +112,16 @@ $model_stat =
...
+  model. For example, { "key" : { "response_stats" : "success" } } indicates the
+  aggregate statistics of successful decoupled responses at "key" for the model,
+  where "key" identifies each decoupled response generated by the model across
+  different requests. For example, given a model that generates three responses,
```
Comment (rmccorm4) [inline docs/protocol/extension_statistics.md:122]: I think it makes sense to call out the example model "decoupled' in this case since it is described to have multiple responses
```suggestion
  different requests. For example, given a decoupled model that generates three responses,
```

File: docs/protocol/extension_statistics.md [hunk 9]
```diff
@@ -111,8 +112,16 @@ $model_stat =
...
+  the key could be "0", "1" and "2" identifying the three responses in order.
+  Check with the model backend on the keys it can return and their meanings.
```
Comment (rmccorm4) [inline docs/protocol/extension_statistics.md:124]: The "keys" aren't backend specific are they? Isn't it a fixed set defined in core/proto?

File: docs/protocol/extension_statistics.md [hunk 10]
```diff
@@ -360,6 +391,12 @@ message ModelStatistics
   // point, the GPU memory usage for models in ONNX Runtime backend and TensorRT
   // backend is usually aligned.
   repeated MemoryUsage memory_usage = 8;
+
+  // The key and value pairs for all decoupled responses statistics. The key is
```
Comment (rmccorm4) [inline docs/protocol/extension_statistics.md:395]:
```suggestion
  // The key and value pairs for all response statistics. The key is
```

File: docs/protocol/extension_statistics.md [hunk 11]
```diff
@@ -428,6 +465,26 @@ message InferStatistics
   StatisticDuration cache_miss = 8;
 }
 
+// Statistics per decoupled response.
```
Comment (rmccorm4) [inline docs/protocol/extension_statistics.md:468]:
```suggestion
// Statistics per response.
```

File: docs/protocol/extension_statistics.md (hunk 7 above)
Comment (kthui) [reply docs/protocol/extension_statistics.md]: Sure, they are all updated on `docs/protocol/extension_statistics.md`.

File: docs/protocol/extension_statistics.md (hunk 9 above)
Comment (kthui) [reply docs/protocol/extension_statistics.md]: They are not backend specific, at this point, unless we allow customizable keys for the backend in the future. I only left the example behind, I think that should be enough.

File: docs/protocol/extension_statistics.md [hunk 12]
```diff
@@ -78,6 +78,7 @@ $model_stat =
   "inference_count" : $number,
   "execution_count" : $number,
   "inference_stats" : $inference_stats,
+  "response_stats" : { $string : $response_stat, ... },
   "batch_stats" : [ $batch_stat, ... ],
```
Comment (rmccorm4) [inline docs/protocol/extension_statistics.md:82]:
```suggestion
  "response_stats" : { $string : $response_stats, ... },
  "batch_stats" : [ $batch_stats, ... ],
```

**Review by rmccorm4 — APPROVED** [review]

**Review by rmccorm4 — APPROVED** [review]
